httpclient/logging_round_tripper.go

```diff
@@ -0,0 +1,150 @@
+func (lm LoggingMode) IsValid() bool {
+	switch lm {
+	case LoggingModeNone, LoggingModeAll, LoggingModeFailed:
+		return true
+	}
+	return false
+}
+
+// LoggingRoundTripper implements http.RoundTripper for logging requests.
+type LoggingRoundTripper struct {
+	// Delegate is the next RoundTripper in the chain.
+	Delegate http.RoundTripper
+
+	// RequestType is a type of request. e.g. service 'auth-service', an action 'login' or specific information to correlate.
+	RequestType string
+
+	// Mode of logging: none, all, failed.
+	Mode LoggingMode
+
+	// SlowRequestThreshold is a threshold for slow requests.
+	SlowRequestThreshold time.Duration
+
+	// LoggerProvider is a function that provides a context-specific logger.
+	// middleware.GetLoggerFromContext is used by default.
+	LoggerProvider func(ctx context.Context) log.FieldLogger
+}
+
+// LoggingRoundTripperOpts represents an options for LoggingRoundTripper.
+type LoggingRoundTripperOpts struct {
+	// RequestType is a type of request. e.g. service 'auth-service', an action 'login' or specific information to correlate.
+	RequestType string
+
+	// Mode of logging: none, all, failed.
+	Mode LoggingMode
+
+	// SlowRequestThreshold is a threshold for slow requests.
+	SlowRequestThreshold time.Duration
+
+	// LoggerProvider is a function that provides a context-specific logger.
+	// middleware.GetLoggerFromContext is used by default.
+	LoggerProvider func(ctx context.Context) log.FieldLogger
+}
+
+// NewLoggingRoundTripper creates an HTTP transport that log requests.
+func NewLoggingRoundTripper(delegate http.RoundTripper) http.RoundTripper {
+	return NewLoggingRoundTripperWithOpts(delegate, LoggingRoundTripperOpts{
+		RequestType: DefaultRequestType,
+	})
+}
+
+// NewLoggingRoundTripperWithOpts creates an HTTP transport that log requests with options.
+func NewLoggingRoundTripperWithOpts(
+	delegate http.RoundTripper, opts LoggingRoundTripperOpts,
+) http.RoundTripper {
+	requestType := opts.RequestType
+	if requestType == "" {
+		requestType = DefaultRequestType
+	}
+
+	return &LoggingRoundTripper{
+		Delegate:             delegate,
+		RequestType:          requestType,
+		Mode:                 opts.Mode,
+		SlowRequestThreshold: opts.SlowRequestThreshold,
+		LoggerProvider:       opts.LoggerProvider,
+	}
+}
+
+// getLogger returns a logger from the context or from the options.
+func (rt *LoggingRoundTripper) getLogger(ctx context.Context) log.FieldLogger {
+	if rt.LoggerProvider != nil {
+		return rt.LoggerProvider(ctx)
+	}
+
+	return middleware.GetLoggerFromContext(ctx)
+}
+
+// RoundTrip adds logging capabilities to the HTTP transport.
+func (rt *LoggingRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	if rt.Mode == LoggingModeNone {
+		return rt.Delegate.RoundTrip(r)
+	}
+
+	ctx := r.Context()
+	logger := rt.getLogger(ctx)
+	start := time.Now()
+
+	resp, err := rt.Delegate.RoundTrip(r)
+	elapsed := time.Since(start)
+	if logger != nil && elapsed >= rt.SlowRequestThreshold {
```

Comment:
The semantics of `LoggingModeFailed` and `LoggingModeAll` seem a bit tricky.
Currently, as I understand, the round tripper only writes logs if `elapsed >= rt.SlowRequestThreshold`. However, in `LoggingModeFailed`, we should log errors even if the request takes less time than `rt.SlowRequestThreshold`.

What do you think about implementing the following logic:
	1.	Log all requests if `LoggingModeAll` is used:
	•	Use the info log level for successful requests.
	•	Use the error log level for failed requests.
	2.	Log failed requests (e.g., connection errors or HTTP status codes <= 400) in `LoggingModeFailed`:
	•	Use the error log level.
	3.	Log slow successful requests in `LoggingModeFailed` if `elapsed >= rt.SlowRequestThreshold`:
	•	Use the warn log level.